### PR TITLE
Plugin/k8s_ingress

### PR DESCRIPTION
**Pull request by networkop**
### 1. Why is this pull request needed and what does it do?
This PR introduces a new `k8s_ingress` plugin to support external resolution of Ingress domain names, similar to what `k8s_external` does for services.

### 2. Which issues (if any) are related?
Issue #2465 

### 3. Which documentation changes (if any) need to be made?
The new plugin readme (part of the PR)
### 4. Does this introduce a backward incompatible change or deprecation?
No, however this would require a new set of RBAC rules, for example:
```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  labels:
  name: excoredns-coredns
rules:
- apiGroups:
  - ""
  resources:
  - endpoints
  - services
  - pods
  - namespaces
  verbs:
  - list
  - watch
- apiGroups:
  - extensions
  resources:
  - ingresses
  verbs:
  - list
  - watch
```

### CODE REVIEW

**Comment by networkop** [comment]
oops, I didn't know there were other CI tests in a separate repo. I'm happy to fix them, once there's a consensus on this PR.

**Comment by chrisohaver** [comment]
Thanks for the PR! I have not done a full review yet.  Initial thoughts:
* There is a significant amount of code duplication with _k8s_external_. So that has me pause to think on how we could do things differently.  For example, how to more easily allow other plugins to "piggy-back" on the kubernetes API connection.
* This appears to cause the _kubernetes_ plugin to maintain a watch on ingresses, even if _k8s_ingress_ is not in use (i.e. affecting all kubernetes users ).  The Kubernetes scalability team is especially sensitive to bumps in memory, and and pressure on the API server.

**Comment by networkop** [comment]
Thanks for your feedback!

> * There is a significant amount of code duplication with _k8s_external_. So that has me pause to think on how we could do things differently.  For example, how to more easily allow other plugins to "piggy-back" on the kubernetes API connection.

This is exactly what I thought as well. Not only that, but to make things work between the two plugins, one has to fallthrough to the next when there's no match found, instead of returning zone apex, like [this](https://github.com/coredns/coredns/pull/3935/files#diff-09e5fe140b7035bc7950cfc93ee6555dR107)


> * This appears to cause the _kubernetes_ plugin to maintain a watch on ingresses, even if _k8s_ingress_ is not in use (i.e. affecting all kubernetes users ).  The Kubernetes scalability team is especially sensitive to bumps in memory, and and pressure on the API server.

That is true, and this can be an argument against merging with k8s_external. 
I'm happy to investigate either option.

**Comment by networkop** [comment]
hi @chrisohaver :wave: 
did you have any further thoughts or feedback on this PR?

**Comment by chrisohaver** [comment]
> did you have any further thoughts or feedback on this PR?

I'll try to carve out time next week to review it in more detail.

File: plugin/k8s_ingress/README.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+				5          ; minimum (4 hours)
+				)
+example.org		5 IN	NS ns1.dns.example.org.
+
+ns1.dns.example.org.  5 IN  A    ....
+ns1.dns.example.org.  5 IN  AAAA ....
+~~~
+
+Note that we use the `dns` subdomain for the records DNS needs (see the `apex` directive). Also
+note the SOA's serial number is static. The IP addresses of the nameserver records are those of the
```
Comment (chrisohaver) [inline plugin/k8s_ingress/README.md:32]: Static serial: This should probably move to a "Known Issues" section... since its not a desired behavior.

File: plugin/k8s_ingress/README.md [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+CoreDNS service.
```
Comment (chrisohaver) [inline plugin/k8s_ingress/README.md:33]: > The IP addresses of the nameserver records are those of the CoreDNS service.

The Cluster IP of the service?  For clients outside fo the cluster, this would be an unreachable IP address.

File: plugin/k8s_ingress/README.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    ttl TTL
+}
+~~~
+
+* **APEX** is the name (DNS label) to use for the apex records; it defaults to `dns`.
+* `ttl` allows you to set a custom **TTL** for responses. The default is 5 (seconds).
+
+## Examples
+
+Enable names under `example.org` to be resolved to in-cluster DNS addresses.
```
Comment (chrisohaver) [inline plugin/k8s_ingress/README.md:61]: Can you rephrase this? It doesn't seem to describe what the example does?

File: plugin/k8s_ingress/README.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+## Name
+
+*k8s_ingress* - resolves ingress FQDNs from outside Kubernetes clusters.
+
+## Description
+
+This plugin allows an additional zone to resolve the external IP address(es) of a Kubernetes
+Ingresses, similar to what `k8s_external` does for services. This plugin is only useful if the *kubernetes* plugin is also loaded.
+
+The plugin uses an external zone to resolve in-cluster IP addresses. It only handles queries for A,
```
Comment (chrisohaver) [inline plugin/k8s_ingress/README.md:12]:
```suggestion
It only handles queries for A,
```

File: plugin/k8s_ingress/README.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
+# k8s_ingress
+
+## Name
+
+*k8s_ingress* - resolves ingress FQDNs from outside Kubernetes clusters.
+
+## Description
+
+This plugin allows an additional zone to resolve the external IP address(es) of a Kubernetes
+Ingresses, similar to what `k8s_external` does for services. This plugin is only useful if the *kubernetes* plugin is also loaded.
```
Comment (chrisohaver) [inline plugin/k8s_ingress/README.md:10]:
```suggestion
This plugin creates records for Kubernetes Ingresses.  This plugin is only useful if the *kubernetes* plugin is also loaded.
```

File: plugin/k8s_ingress/external.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
+/*
+Package ingress implements ingress names for kubernetes clusters.
+
+This plugin only handles three qtypes (except the apex queries, because those are handled
+differently). We support A, AAAA and SRV request, for all other types we return NODATA or
```
Comment (chrisohaver) [inline plugin/k8s_ingress/external.go:5]: SRV records don't appear to be supported in this PR.

File: plugin/k8s_ingress/external.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+NXDOMAIN depending on the state of the cluster.
+
+A plugin willing to provide these services must implement the Externaler interface, although it
```
Comment (chrisohaver) [inline plugin/k8s_ingress/external.go:8]: Externaler -> Ingresser

File: plugin/k8s_ingress/external.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+	"github.com/coredns/coredns/plugin/etcd/msg"
+	"github.com/coredns/coredns/plugin/pkg/upstream"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+// Ingresser defines the interface that a plugin should implement in order to be used by External.
+type Ingresser interface {
+	// External returns a slice of msg.Services that are looked up in the backend and match
```
Comment (chrisohaver) [inline plugin/k8s_ingress/external.go:27]: Copy-pasted comments don't match property names in this interface declaration.

File: plugin/k8s_ingress/external.go [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+func New() *Ingress {
+	e := &Ingress{hostmaster: "hostmaster", ttl: 5, apex: "dns"}
+	return e
+}
+
+// ServeDNS implements the plugin.Handle interface.
+func (e *Ingress) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+
+	//log.Printf("Service query %s", state.Name())
```
Comment (chrisohaver) [inline plugin/k8s_ingress/external.go:59]: These commented-out debugs should be removed before merging.

File: plugin/k8s_ingress/external.go [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+	switch state.QType() {
+	case dns.TypeA:
+		m.Answer = e.a(ctx, svc, state)
+	case dns.TypeAAAA:
+		m.Answer = e.aaaa(ctx, svc, state)
+	default:
+		m.Ns = []dns.RR{e.soa(state)}
+	}
+
+	// If we did have records, we have to continue to let other plugins pick it up.
```
Comment (chrisohaver) [inline plugin/k8s_ingress/external.go:105]: missing "not" ? "If we did _not_ have records," ...

This is implicit "fallthrough" behavior, which we should avoid I think.  Instead, add a "fallthough" option to the plugin, and continue down the plugin chain only if fallthough is enabled (see other plugins: kubernetes, hosts).  Also, to be consistent with other fallthrough implementations, we should only be falling through when the result is NXDOMAIN.  e.g. if someone does a TXT query on name that exists, we should return NODATA instead of allowing another plugin to answer.

File: plugin/k8s_ingress/external.go [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+		return plugin.NextOrFailure(e.Name(), e.Next, ctx, w, r)
+	}
+
+	if e.externalFunc == nil {
+		return plugin.NextOrFailure(e.Name(), e.Next, ctx, w, r)
+	}
+	//log.Printf("Found matching zone %s", zone)
+	state.Zone = zone
+	for _, z := range e.Zones {
+		// TODO(miek): save this in the External struct.
```
Comment (chrisohaver) [inline plugin/k8s_ingress/external.go:71]: I assume this TODO is copy-pasted?  Does it still apply in this context?

File: plugin/k8s_ingress/external.go [hunk 12]
```diff
@@ -0,0 +1,115 @@
+/*
```
Comment (chrisohaver) [inline plugin/k8s_ingress/external.go:1]: Should this file be named `ingress.go`?  is `external.go` a remanent of k8s_external?

File: plugin/k8s_ingress/msg_to_dns.go [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+				rr := s.NewAAAA(state.QName(), ip)
+				rr.Hdr.Ttl = e.ttl
+				records = append(records, rr)
+			}
+		}
+	}
+	return records
+}
+
+// not sure if this is even needed.
```
Comment (chrisohaver) [inline plugin/k8s_ingress/msg_to_dns.go:81]: Please find out, and remove if not needed.

File: plugin/kubernetes/controller.go [hunk 14]
```diff
@@ -116,6 +124,17 @@ func newdnsController(ctx context.Context, kubeClient kubernetes.Interface, opts
 		object.DefaultProcessor(object.ToService(opts.skipAPIObjectsCleanup)),
 	)
 
+	dns.ingLister, dns.ingController = object.NewIndexerInformer(
```
Comment (chrisohaver) [inline plugin/kubernetes/controller.go:127]: This informer must be conditionally created, like the pod informer.
When k8s_ingress is not in use, kubernetes should not establish a watch on ingress objects.

File: plugin/k8s_ingress/README.md [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+The plugin uses an external zone to resolve in-cluster IP addresses. It only handles queries for A,
+AAAA; all others result in NODATA responses. To make it a proper DNS zone, it handles SOA and NS queries for the apex of the zone.
```
Comment (chrisohaver) [inline plugin/k8s_ingress/README.md:13]: Even though it's implied, probably should call out specifically that PTR requests (reverse lookups) are not handled.

File: plugin/kubernetes/controller.go [hunk 16]
```diff
@@ -21,13 +22,18 @@ import (
...
+	ingHostIndex          = "IngressHost"
 	epNameNamespaceIndex  = "EndpointNameNamespace"
 	epIPIndex             = "EndpointsIP"
 )
 
 type dnsController interface {
 	ServiceList() []*object.Service
+	IngressList() []*object.Ingress
+	IngIndex(string) []*object.Ingress
+	IngIndexReverse(string) []*object.Ingress
```
Comment (chrisohaver) [inline plugin/kubernetes/controller.go:36]: This isn't ideal - that we have to modify the controller interface like this.
Perhaps there is another approach...

**Review by chrisohaver — CHANGES_REQUESTED** [review]
@networkop, I spent a few hours going over this.  IMO there is too much code (and comment) duplication for this to be feasibly included as an internal plugin.

**Comment by chrisohaver** [comment]
An alternate approach could be to establish a separate connection and watch on the API.  I think the k8s client-go consolidates connections to minimize overhead.

**Comment by miekg** [comment]
please send a PR to coredns/coredns.io where this can live as an external plugin. ta.

**Comment by networkop** [comment]
@chrisohaver thanks for taking time to review this. I also felt there was too much code duplication and changes to other modules/plugins. This was a "naive" implementation where I tried to follow the current plugin (k8s_external) as much as possible. The goal was to open a discussion and get your opinion about what needs to change. So the main two issues with the logic (excluding comments/readme) are:
1. Modifying the k8s controller interface
2. Wrong fallthrough behaviour

I will try to refactor the code to make sure all k8s controller code is pulled into the ingress plugin itself, so it becomes more self-contained. 
However, this still would require changes to RBAC and, consequently, helm charts. So if you think that there's no point in continuing with this PR here, I'm happy work on it externally and, as @miekg suggested, simply list it as one of the external plugins.

**Comment by chrisohaver** [comment]
The fallthrough behavior is easily fixable.  Avoiding the interface changes less so.  

The k8s controller code is somewhat complex due to a requirement of us keeping under a target memory limit set by Kubernetes.  E.g. we use customized indexer informers and do object conversions. Copying that structure verbatim would carry a lot of bulk and complexity you don't need - you should be able to use the default informers included in client-go.

Another other issue is the "always-on" watch on ingresses, even if k8s_ingress is not enabled in the corefile.
With that always on,  the RBAC rules would need to be updated.  However, if ingresses are only watched if k8s_ingress is enabled, then the RBAC rules only need to permit ingress list/watch when k8s_ingress is enabled.  Since k8s_ingress would not be enabled by default, the RBAC rules don't need to allow it by default.  If a user enables k8s_ingress, they would need to edit the RBAC rules.  This could be described in the README.  In the helm chart, if k8s_ingress is added as a configurable option, then it would need to adjust RBAC rules accordingly.

I think for now this should be an external plugin, and refactored to not require changes to the kubernetes plugin controller interface.

In the future, it would be nice for kubernetes plugin to make the informers/indexers generally available to other plugins (i.e. a way for other plugins to add informer/indexers to the controller or access existing ones).  If we work that out, and k8s_ingress proves to be a popular feature, we can consider integrating it into the tree as an internal plugin.

**Comment by networkop** [comment]
Understood. I'll close this PR then. I'll update the issue #2465 if/when the external plugin is ready.
